Tests/SourceKitLSPTests/WorkspaceTests.swift

```diff
@@ -967,4 +968,171 @@ final class WorkspaceTests: XCTestCase {
+    let sdkArgs =
+      if let defaultSDKPath {
+        """
+        -sdk '\(defaultSDKPath)'
+        """
+      } else {
+        ""
+      }
+
+    let project = try await MultiFileTestProject(files: [
+      ".sourcekit-lsp/config.json": """
+      {
+        "defaultWorkspaceType": "compilationDatabase"
+      }
+      """,
+      "src/Foo.swift": """
+      #if HAVE_SETTINGS
+      #error("Have settings")
+      #endif
+      """,
+      "Sources/MyLib/Bar.swift": "",
+      "build/compile_commands.json": """
+      [
+      {
+        "directory": "$TEST_DIR",
+        "command": "swiftc $TEST_DIR/src/Foo.swift -DHAVE_SETTINGS \(sdkArgs)",
+        "file": "src/Foo.swift",
+        "output": "$TEST_DIR/build/Foo.swift.o"
+      },
+      ]
+      """,
+      "Package.swift": """
+      // swift-tools-version: 5.7
+
+      import PackageDescription
+
+      let package = Package(
+        name: "MyLib",
+        targets: [
+          .target(name: "MyLib"),
+        ]
+      )
+      """,
+    ])
+    let (uri, _) = try project.openDocument("Foo.swift")
+    let diagnostics = try await project.testClient.send(
+      DocumentDiagnosticsRequest(textDocument: TextDocumentIdentifier(uri))
+    )
+    XCTAssertEqual(
+      diagnostics.fullReport?.items.map(\.message),
+      ["Have settings"]
+    )
+  }
+
+  func testImplicitWorkspaceOptionsOverrideBuildSystem() async throws {
+    let sdkArgs =
+      if let defaultSDKPath {
+        """
+        -sdk '\(defaultSDKPath)'
+        """
+      } else {
+        ""
+      }
+
+    let project = try await MultiFileTestProject(files: [
+      "projA/.sourcekit-lsp/config.json": """
+      {
+        "defaultWorkspaceType": "compilationDatabase"
+      }
+      """,
+      "projA/src/Foo.swift": """
+      #if HAVE_SETTINGS
+      #error("Have settings")
+      #endif
+      """,
+      "projA/Sources/MyLib/Bar.swift": "",
+      "projA/build/compile_commands.json": """
+      [
+      {
+        "directory": "$TEST_DIR/projA",
+        "command": "swiftc $TEST_DIR/projA/src/Foo.swift -DHAVE_SETTINGS \(sdkArgs)",
+        "file": "src/Foo.swift",
+        "output": "$TEST_DIR/projA/build/Foo.swift.o"
+      },
+      ]
+      """,
+      "projA/Package.swift": """
+      // swift-tools-version: 5.7
+
+      import PackageDescription
+
+      let package = Package(
+        name: "MyLib",
+        targets: [
+          .target(name: "MyLib"),
+        ]
+      )
+      """,
+    ])
+    let (uri, _) = try project.openDocument("Foo.swift")
+    let diagnostics = try await project.testClient.send(
+      DocumentDiagnosticsRequest(textDocument: TextDocumentIdentifier(uri))
+    )
+    XCTAssertEqual(
+      diagnostics.fullReport?.items.map(\.message),
+      ["Have settings"]
+    )
+  }
+
+  func testWorkspaceOptionsCanAddSearchPaths() async throws {
+    let sdkArgs =
+      if let defaultSDKPath {
+        """
+        -sdk '\(defaultSDKPath)'
+        """
+      } else {
+        ""
+      }
+
```

Comment:
Might be worth sharing this between test cases, similar to what we do in `BuildServerTestProject`. Also, we should replace `\` with `\\` in `defaultSDKPath`, so this doesn’t fail on Windows.